src/snowflake/cli/api/connections.py

```diff
@@ -0,0 +1,206 @@
+            self.connection_name = get_default_connection_name()
+
+    def as_nonempty_dict(self) -> dict:
+        return {k: v for (k, v) in asdict(self).items() if v is not None}
+
+    def build_connection(self):
+        from snowflake.cli._app.snow_connector import connect_to_snowflake
+
+        # Ignore warnings about bad owner or permissions on Windows
+        # Telemetry omit our warning filter from config.py
+        if IS_WINDOWS:
+            warnings.filterwarnings(
+                action="ignore",
+                message="Bad owner or permissions.*",
+                module="snowflake.connector.config_manager",
+            )
+
+        return connect_to_snowflake(**self.as_nonempty_dict())
+
+
+class OpenConnectionCache:
+    """
+    A connection cache that transparently manages SnowflakeConnection objects
+    and is keyed by ConnectionContext objects, e.g. cache[ctx].execute_string(...).
+    Connections are automatically closed after CONNECTION_CLEANUP_SEC, but
+    are guaranteed to be open (if config is valid) when returned by the cache.
+    """
+
+    connections: dict[str, SnowflakeConnection]
+    cleanup_futures: dict[str, asyncio.TimerHandle]
+
+    CONNECTION_CLEANUP_SEC: float = 10.0 * 60
+    """Connections are closed this many seconds after the last time they are accessed."""
+
+    def __init__(self):
+        self.connections = {}
+        self.cleanup_futures = {}
+
+    def __getitem__(self, ctx):
+        if not isinstance(ctx, ConnectionContext):
+            raise ValueError(
+                f"Expected key to be ConnectionContext but got {repr(ctx)}"
+            )
+        key = repr(ctx)
+        if not self._has_open_connection(key):
+            self._insert(key, ctx)
+        self._touch(key)
+        return self.connections[key]
+
+    def clear(self):
+        """Closes all connections and resets the cache to its initial state."""
+        connection_keys = list(self.connections.keys())
+        for key in connection_keys:
+            self._cleanup(key)
+
+        # if any orphaned futures still exist, clean them up too
+        for key in self.cleanup_futures:
+            self.cleanup_futures[key].cancel()
+        self.cleanup_futures.clear()
+
+    def _has_open_connection(self, key: str):
+        return key in self.connections
+
+    def _insert(self, key: str, ctx: ConnectionContext):
+        try:
+            self.connections[key] = ctx.build_connection()
+        except Exception:
+            logger.debug("ConnectionCache: failed to connect using {key}; not caching.")
+            raise
+
+    def _cancel_cleanup_future_if_exists(self, key: str):
+        if key in self.cleanup_futures:
+            self.cleanup_futures.pop(key).cancel()
+
+    def _touch(self, key: str):
+        """
+        Extend the lifetime of the cached connection at the given key.
+        """
+        try:
+            loop = asyncio.get_event_loop()
+        except (RuntimeError, DeprecationWarning):
```

Comment:
`DeprecationWarning`s aren't thrown; on earlier python versions this just returns `Non` if there's no loop, so we need to check for that before calling `.call_later()`